### PR TITLE
Add ccm cluster logging facility

### LINKED ISSUE #214: Add ccm logging
During testing it would be helpful to have ccm debug logging in addition to dtest logging (that's saved in `dtest.log` artifact).

@fruch:
> CCM as a library should use logging with specific logger namespace ccm.*
> 
> And then we could configure that logger in dtest to write to the stdout as needed.

### PR DESCRIPTION
**Pull request by bhalevy**
Add a logging facility to Cluster class.
Provide optional `ccm_logger` and `ccm_log_level` parameters provided for integration with dtest logging.
See https://github.com/bhalevy/scylla-dtest/commit/9dac036a8e8ff5766559f3d158b872972cee4bec for planned usage.

Fixes #214

Note: this series depends on https://github.com/scylladb/scylla-ccm/pull/253
that was recently merged, hence it's based off of `scylla-ccm/next`.

Output example:
```
(env) $ PRINT_DEBUG=true CASSANDRA_DIR=../scylla/build/debug NOSE_PROCESSES=2 SCYLLA_EXT_OPTS="--smp 2" nosetests -s -v --process-timeout=7200 cfid_test:TestCFID.cfid_test
2020-10-18 10:55:35.257537 going to run tests in parallel
2020-10-18 10:55:35.257622 using the RandomClusterIdAllocator
2020-10-18 10:55:35.284624 cfid_test.py:TestCFID.cfid_test - cluster ccm directory: /local/home/bhalevy/.dtest/dtest-i5u1fsh5
2020-10-18 10:55:35.284784 cfid_test.py:TestCFID.cfid_test - Starting Scylla cluster from directory ../scylla/build/debug
2020-10-18 10:55:35.288229 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: Started cluster version 3.0 installed in /local/home/bhalevy/dev/scylla
2020-10-18 10:55:35.289910 cfid_test.py:TestCFID.cfid_test - Allocated cluster ID 48: /local/home/bhalevy/.dtest/dtest-i5u1fsh5
2020-10-18 10:55:35.344042 cfid_test.py:TestCFID.cfid_test - Scylla mode is 'debug'
2020-10-18 10:55:35.344126 cfid_test.py:TestCFID.cfid_test - Cluster *_request_timeout_in_ms=30000, cql request_timeout=90
2020-10-18 10:55:35.345772 cfid_test.py:TestCFID.cfid_test - configuring skip_wait_for_gossip_to_settle=0 for single_node test
2020-10-18 10:55:35.421944 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Starting scylla: args=['/local/home/bhalevy/.dtest/dtest-i5u1fsh5/test/node1/bin/scylla', '--options-file', '/local/home/bhalevy/.dtest/dtest-i5u1fsh5/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.48.1', '--collectd-hostname', 'localhost.localdomain.node1', '--smp', '2', '--developer-mode', 'true', '--memory', '1024M', '--default-log-level', 'info', '--blocked-reactor-notify-ms', '5000', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.48.1', '--unsafe-bypass-fsync', '1'] wait_other_notice=False wait_for_binary_proto=False
2020-10-18 10:55:56.256866 cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 10:55:57.270330 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool flush
2020-10-18 10:55:59.953793 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool cfstats timeout=60
2020-10-18 10:56:10.265263 cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 10:56:11.494248 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool flush
2020-10-18 10:56:12.914373 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool cfstats timeout=60
2020-10-18 10:56:21.581253 cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 10:56:22.628048 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool flush
2020-10-18 10:56:23.994833 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool cfstats timeout=60
2020-10-18 10:56:32.298134 cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 10:56:33.325099 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool flush
2020-10-18 10:56:35.007422 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool cfstats timeout=60
2020-10-18 10:56:43.001427 cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 10:56:43.991935 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool flush
2020-10-18 10:56:45.308304 cfid_test.py:TestCFID.cfid_test - ccm: cluster test: node1: Running nodetool cfstats timeout=60
2020-10-18 10:56:53.356763 cfid_test.py:TestCFID.cfid_test - stopping ccm cluster test at: /local/home/bhalevy/.dtest/dtest-i5u1fsh5
2020-10-18 10:56:54.408121 cfid_test.py:TestCFID.cfid_test - Freeing cluster ID 48: link /local/home/bhalevy/.dtest/48
cfid_test (cfid_test.TestCFID) ... ok

----------------------------------------------------------------------
Ran 1 test in 79.146s

OK
```

### CODE REVIEW

File: ccmlib/cluster.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 class Cluster(object):
 
-    def __init__(self, path, name, partitioner=None, install_dir=None, create_directory=True, version=None, verbose=False, snitch='org.apache.cassandra.locator.PropertyFileSnitch', **kwargs):
+    def __init__(self, path, name, partitioner=None, install_dir=None, create_directory=True, version=None, verbose=False, snitch='org.apache.cassandra.locator.PropertyFileSnitch', ccm_logger=None, ccm_log_level=None, **kwargs):
```
Comment (fruch) [inline ccmlib/cluster.py:20]: why not just use `logging` module directly ?, to retain the print option in dtest ?

Comment (fruch) [reply ccmlib/cluster.py:20]: i.e. something like
```python
logger = logging.getLogger("scylla.ccm")
logger.info("whatever")
```

and then on dtest configure what's  needed for it to be printed on stdout, when we need

Not only the cluster class has logging we care about, also other part ccm might have logs we want utility functions, the class download the version and such.

**Comment by bhalevy** [comment]
With 04efaf0 ccm logging reaches `dtest.log` (but not printed to the stdout) and looks as follows:
```
2020-10-18 17:13:24,715 843356  dtest                          DEBUG    | going to run tests in parallel
2020-10-18 17:13:24,715 843356  dtest                          DEBUG    | using the RandomClusterIdAllocator
2020-10-18 17:13:24,744 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - cluster ccm directory: /local/home/bhalevy/.dtest/dtest-ti8mcnsq
2020-10-18 17:13:24,744 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - Starting Scylla cluster from directory ../scylla/build/debug
2020-10-18 17:13:24,748 843379  ccm                            DEBUG    | ccm: cluster test: Started cluster version 4.3.dev installed in /local/home/bhalevy/dev/scylla
2020-10-18 17:13:24,750 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - Allocated cluster ID 32: /local/home/bhalevy/.dtest/dtest-ti8mcnsq
2020-10-18 17:13:24,807 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - Scylla mode is 'debug'
2020-10-18 17:13:24,807 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - Cluster *_request_timeout_in_ms=30000, cql request_timeout=90
2020-10-18 17:13:24,809 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - configuring skip_wait_for_gossip_to_settle=0 for single_node test
2020-10-18 17:13:24,877 843379  ccm                            DEBUG    | ccm: cluster test: node1: Starting scylla: args=['/local/home/bhalevy/.dtest/dtest-ti8mcnsq/test/node1/bin/scylla', '--options-file', '/local/home/bhalevy/.dtest/dtest-ti8mcnsq/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.32.1', '--collectd-hostname', 'localhost.localdomain.node1', '--smp', '2', '--developer-mode', 'true', '--memory', '1024M', '--default-log-level', 'info', '--blocked-reactor-notify-ms', '5000', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.32.1', '--unsafe-bypass-fsync', '1'] wait_other_notice=False wait_for_binary_proto=False
2020-10-18 17:13:45,297 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 17:13:46,330 843379  ccm                            DEBUG    | ccm: cluster test: node1: Running nodetool flush
2020-10-18 17:13:48,752 843379  ccm                            DEBUG    | ccm: cluster test: node1: Running nodetool cfstats timeout=60
2020-10-18 17:13:59,063 843379  dtest                          DEBUG    | cfid_test.py:TestCFID.cfid_test - CREATE COLUMNFAMILY cf (key int PRIMARY KEY, c1 int) WITH comment='test cf' AND compression = {} AND gc_grace_seconds=0
2020-10-18 17:14:00,142 843379  ccm                            DEBUG    | ccm: cluster test: node1: Running nodetool flush
2020-10-18 17:14:01,577 843379  ccm                            DEBUG    | ccm: cluster test: node1: Running nodetool cfstats timeout=60
```

I think this is a significant addition on its own and worth merging.

File: ccmlib/common.py [hunk 2]
```diff
@@ -39,6 +40,14 @@
 CONFIG_FILE = "config"
 CCM_CONFIG_DIR = "CCM_CONFIG_DIR"
 
+logger = logging.getLogger('ccm')
+ccm_log_level = os.environ.get('CCM_LOG_LEVEL', None)
```
Comment (fruch) [inline ccmlib/common.py:44]: I think all the level control can be handled only in dtest.

and should maybe part of the command line options for when ccm is called directly

File: ccmlib/common.py [hunk 3]
```diff
@@ -39,6 +40,14 @@
...
+if ccm_log_level is not None:
+    ccm_log_level = ccm_log_level.upper()
+elif os.environ.get('PRINT_DEBUG', '').lower() in ('yes', 'true'):
```
Comment (fruch) [inline ccmlib/common.py:47]: `PRINT_DEBUG`  here is pointless, and doesn't actually does what it suggest like in dtest, I think it should be removed from here.

**Review by fruch — CHANGES_REQUESTED** [review]
Control of log leveling should be in global place like that.
It should be handled as much as possible by the called of ccm (if dtest, or the command line ccm)
library just define the name, and then you can control the level and other stuff, configuration should be in one place, and not spitted across two different code base, it would sometime can conflict and make the config harder

File: ccmlib/common.py (hunk 2 above)
Comment (bhalevy) [reply ccmlib/common.py:44]: ccm can also run as a standalone program

**Comment by bhalevy** [comment]
> Control of log leveling should be in global place like that.
> It should be handled as much as possible by the called of ccm (if dtest, or the command line ccm)
> library just define the name, and then you can control the level and other stuff, configuration should be in one place, and not spitted across two different code base, it would sometime can conflict and make the config harder

okay, I'll take that out

**Comment by bhalevy** [comment]
@fruch ping re-review

**Review by fruch — APPROVED** [review]

**Pull request merged**